### PR TITLE
[Typed] ReplicatedEntity API

### PR DESCRIPTION
**Pull request by negokaz**
related issue: #44

### CODE REVIEW

File: src/main/scala/lerna/akka/entityreplication/typed/ClusterReplication.scala [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package lerna.akka.entityreplication.typed
+
+import akka.actor.typed.{ ActorRef, ActorSystem, Extension, ExtensionId }
+
+object ClusterReplication extends ExtensionId[ClusterReplication] {
```
Comment (negokaz) [inline src/main/scala/lerna/akka/entityreplication/typed/ClusterReplication.scala:5]: [memo] see also: [akka/ClusterSharding.scala at v2.6.13 · akka/akka](https://github.com/akka/akka/blob/v2.6.13/akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/scaladsl/ClusterSharding.scala)

File: src/main/scala/lerna/akka/entityreplication/typed/ReplicatedEntity.scala [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    */
+  def apply[M](typeKey: ReplicatedEntityTypeKey[M])(
+      createBehavior: ReplicatedEntityContext[M] => Behavior[M],
+  ): ReplicatedEntity[M, ReplicationEnvelope[M]] = ???
+}
+
+/**
+  * Defines how the entity should be created. Used in [[ClusterReplication.init()]].
+  */
+trait ReplicatedEntity[M, E] {
```
Comment (negokaz) [inline src/main/scala/lerna/akka/entityreplication/typed/ReplicatedEntity.scala:18]: [memo] see also: 

> ```
> final class Entity[M, E] private[akka] (
> ```
> **[akka/ClusterSharding.scala at v2.6.13 · akka/akka](https://github.com/akka/akka/blob/v2.6.13/akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/scaladsl/ClusterSharding.scala#L245)**

File: src/main/scala/lerna/akka/entityreplication/typed/ReplicatedEntityContext.scala [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package lerna.akka.entityreplication.typed
+
+import akka.actor.typed.ActorRef
+
+class ReplicatedEntityContext[M](
```
Comment (negokaz) [inline src/main/scala/lerna/akka/entityreplication/typed/ReplicatedEntityContext.scala:5]: [memo] see also:

> ```
> final class EntityContext[M](
> ```
> **[akka/ClusterSharding.scala at v2.6.13 · akka/akka](https://github.com/akka/akka/blob/v2.6.13/akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/scaladsl/ClusterSharding.scala#L353)**

File: src/main/scala/lerna/akka/entityreplication/typed/ReplicatedEntityTypeKey.scala [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  /**
+    * Creates an [[ReplicatedEntityTypeKey]]. The `name` must be unique.
+    */
+  def apply[M](name: String)(implicit mTag: ClassTag[M]): ReplicatedEntityTypeKey[M] = ???
+}
+
+/**
+  * The key of an entity type, the `name` must be unique.
+  */
+trait ReplicatedEntityTypeKey[-M] {
```
Comment (negokaz) [inline src/main/scala/lerna/akka/entityreplication/typed/ReplicatedEntityTypeKey.scala:16]: [memo] see also: 

> ```
> @DoNotInherit trait EntityTypeKey[-T] {
> ```
> **[akka/ClusterSharding.scala at v2.6.13 · akka/akka](https://github.com/akka/akka/blob/v2.6.13/akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/scaladsl/ClusterSharding.scala#L387)**

File: src/main/scala/lerna/akka/entityreplication/typed/ReplicationEnvelope.scala [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package lerna.akka.entityreplication.typed
+
+import akka.actor.WrappedMessage
+
+/**
+  * Envelope type that is used by with Cluster Replication
+  */
+final case class ReplicationEnvelope[M](entityId: String, message: M) extends WrappedMessage
```
Comment (negokaz) [inline src/main/scala/lerna/akka/entityreplication/typed/ReplicationEnvelope.scala:8]: [memo] see also:

> ```
> final case class ShardingEnvelope[M](entityId: String, message: M) extends WrappedMessage {
> ```
> **[akka/ShardingMessageExtractor.scala at v2.6.13 · akka/akka](https://github.com/akka/akka/blob/v2.6.13/akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ShardingMessageExtractor.scala#L113)**

File: src/main/scala/lerna/akka/entityreplication/typed/ClusterReplication.scala [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+/**
+  * This extension provides fast recovery by creating replicas of entities in multiple locations
+  * and always synchronizing their status.
+  */
+trait ClusterReplication extends Extension {
+
+  def init[M, E](entity: ReplicatedEntity[M, E]): ActorRef[E]
+}
```
Comment (negokaz) [inline src/main/scala/lerna/akka/entityreplication/typed/ClusterReplication.scala:19]: [memo]

`entityRefFor` will be defined after `ReplicatedEntityRef` is defined.

**Review by tksugimoto — APPROVED** [review]

**Pull request merged**